rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+
+### CHANGELOG
+
+`feat(synth): custom resolvers`
+
+### README
+
+Custom resolvers are a mechanism to inject custom logic into the cdk8s value resolution process. 
+It allows to transform any value just before being written to the Kubernetes manifest.
+
+To define a custom resolver, first create a class that implements the `ITokenResolver` interface:
+
+```ts
+import { ITokenResolver, ResolutionContext } from 'cdk8s';
+
+export class MyCustomResolver implements ITokenResolver {
+
+  public resolve(context: ResolutionContext, value: any): any {
+    // run some custom logic
+  }
+
+}
+```
+
+The `context` argument contains information about the value that is currently being resolved:
+
+- **obj**: `ApiObject` currently being resolved.
+- **key**: Array containing the JSON path elements of they keys leading up to the value.
+
+When you create a cdk8s `Chart`, pass the resolver instance to it via the `resolver` property:
+
+```ts
+import { App, Chart } from 'cdk8s'
+
+const app = new App();
+new Chart(app, 'Chart', { resolver: new MyCustomResolver() });
+```
+
+When you run `cdk8s synth`, your custom logic will be invoked, and the return 
+value will replace the original value and be written to the manifest. For example, if 
+you define a Kubernetes service like so:
+
+```ts
+new KubeService(this, 'Service', {
+  spec: {
+    type: 'LoadBalancer',
+  }
+});
+```
+
+Your resolver will be invoked with the following arguments:
+
+- **context**
+  - *obj*: The `KubeService` instance of type `ApiObject`.
+  - *key*: `['spec', 'type']`
+- **value**: `LoadBalancer`
+
+
+One common use-case for this feature is to automatically resolve deploy time 
+attributes of cloud resources, and pass them to your Kubernetes workloads. To that end, 
+cdk8s provides a resolver for the following CDK frameworks:
+
+#### AWS Cloud Development Kit
+
+In this example, we create an S3 `Bucket` with the AWS CDK, and pass its (deploy time generated) name 
+as an environment variable to a Kubernetes `CronJob` resource.
+
+```ts
+import * as aws from 'aws-cdk-lib';
+import * as k8s from 'cdk8s';
+import * as kplus from 'cdk8s-plus-26';
+
+import { AwsCdkTokenResolver } from '@cdk8s/aws-cdk-token-resolver';
+
+const awsApp = new aws.App();
+const stack = new aws.Stack(awsApp, 'Stack');
+
+const resolver = new AwsCdkTokenResolver(stack);
+
+const k8sApp = new k8s.App();
+const manifest = new k8s.Chart(k8sApp, 'Manifest', { resolver });
+
+const bucket = new aws.aws_s3.Bucket(stack, 'Bucket');
+
+new kplus.CronJob(manifest, 'CronJob', {
+  schedule: k8s.Cron.daily(),
+  containers: [{
+    image: 'job',
+    envVariables: {
+      // passing the bucket name via an env variable
+      BUCKET_NAME: kplus.EnvValue.fromValue(bucket.bucketName),
+    }
+ }]
+});
+
+awsApp.synth();
+k8sApp.synth();
+```
+
+Notice we create two applications: one for our cdk8s constructs, and one for our AWS CDK constructs.
+Both are defined and synthesized in the same file, but can be separated as needed.
```

Comment:
Not too far though, it will have to be the same application at runtime.

Reply:
Revised